### ISSUE TITLE
Implement type DOUBLE

### ISSUE THREAD

**Issue by SamProtas** [balderdashy/waterline#1067]
type: double is currently cast as 'string' as it's not officially supported.

https://github.com/balderdashy/waterline/blob/317cdf0173990face8afe67b336bc6e6d0fe5630/lib/waterline/core/typecast.js#L40

Many adapters implement double (sails-postgresql) or at least degrade gracefully (sails-mysql coerces to FLOAT).

Please add DOUBLE to types.js in order to allow adapters to deal with type DOUBLE in the best way for that database.

**Comment by SamProtas** [comment]
PR submitted:

https://github.com/balderdashy/waterline/pull/1068

**Comment by devinivy** [comment]
Hey @SamProtas, does this feature in the works look like it might do the trick? https://github.com/balderdashy/waterline/pull/981

**Comment by SamProtas** [comment]
@devinivy I haven't delved to deep but that looks like a broader (inclusive) feature/fix. Is there any timeline for the implementation of #981?

My only plug for #1088 as a bandaid is that *double* is an extremely common field type that's already implemented in many (possibly all) of the most common adapters. It's a non-breaking change that we could merge today :+1: 

These 3 lines of code will given everyone access to type=double today!

I get if bandaids are frowned upon though....

**Comment by devinivy** [comment]
@dmarcelino what do we know about the timeline on #981?  I do understand the desire have more granular typing.  To get your PR through we will have to make sure all the supported adapters do actually understand how to deal with that type.  Does javascript have enough bits per float to support what most DBs would consider a double?

**Comment by dmarcelino** [comment]
> what do we know about the timeline on #981?

Still waiting on @particlebanana's input. Not sure if he had a chance to look at it. 

>  Does javascript have enough bits per float to support what most DBs would consider a double?

[javascript numbers](http://www.w3schools.com/js/js_numbers.asp) are 64-bit Floating Point so I think it would align well with DB `double`.

> I get if bandaids are frowned upon though....

I think this would open a worse precedent. Supporting `double` in waterline core would suggest all adapters can deal with it which I doubt is the case. I would prefer pushing  #981 as it's a definitive solution for the different types issues.

**Comment by SamProtas** [comment]
Full disclosure, I haven't tested this yet. 

Two wrongs don't necessarily make a right, but to play devil's advocate, anchor doesn't appear to be able to handle `type: 'objectid'` and I don't see anything in waterline to deal with this either....

That said, +1 for #981

**Comment by SamProtas** [comment]
@dmarcelino 

sails-postgresql fails with `Unregistered type given: objectid` when using the officially supported type objectid.

Separate issue?

**Comment by awm33** [comment]
@dmarcelino Wouldn't the adapters that don't support double end up rounding javascript numbers anyway? All javascript numbers are 64 bit floating point precision numbers. So we actually have a type mismatch. AFAIK nothing rounds the number to 32 bit before it get's passed to the adapter, so float can be a misleading type

**Comment by devinivy** [comment]
@SamProtas this is legacy code to deal with mongo.  It's non-ideal, but it was a design decision of waterline in the fairly distant past, as it transitioned from sql-only to entirely adapter-based (including very importantly, as you can imagine, the nosql option of mongo).

@awm33 Is there an issue of having extra precision while performing calculations with a value that came out of the database?  Wont it still come out of the database with 32 bits of precision and 32 unused bits?  That is, if you were to retrieve a value from the db, commit that value into the db, then retrieve it again, you'd get the same number you started with.  Of course when using certain db types it's implied that it should be convertible to some "equivalent" type in javascript.  If it's a priority of the community to support exact floating point precision for some reason, I'm sure some day we'll see a complex integration with big-integer or double-bits, etc.

The main takeaway– think of the type you put in your model definition as being a javascript type that you'll _receive_, and not the type that is _stored_.  Right now waterline mostly conflates those two ideas, and #981 is the beginning of the road to righting that issue.

**Comment by dmarcelino** [comment]
> sails-postgresql fails with Unregistered type given: objectid when using the officially supported type objectid.

@SamProtas, I believe `objectid` was a sort of a hack for sails-mongo, one that I don't know the full story, @particlebanana? Anyway it was done before PR #981 so there wasn't any proposal on how to separate core types from adapter/DB types.

> Separate issue?

Yes, it is a separate issue and one that will have to happen in a major release as changing it will be a breaking change.

> AFAIK nothing rounds the number to 32 bit before it get's passed to the adapter, so float can be a misleading type

@awm33, if an adapter tries to define a field with `type: double` and doesn't support it will likely throw an error or treat it as a `string`.

My long term view for waterline `type` is that it should correlate strongly with javascript types. The `dbType` would allow explicitly telling the adapter to what the value should be converted to. If the adapter does not support the specific `dbType` it can default to whatever handles the javascript `type` better. There is a very good discussion about this in https://github.com/balderdashy/waterline/pull/977#issuecomment-96756804, the original custom type PR.

**Comment by awm33** [comment]
@devinivy Right. I guess I was thinking before the point of storage, or in my case, we're syncing values betwen us and a 3rd party API which uses double precision, in JSON

@dmarcelino I agree with making waterline types as close to javascript as possible. Particularly because a lot of people end up serving up JSON with the results at some point. But, that makes me think double precision should be the default, or even just "number", considering there is the 2^53 barrier with integers. I suppose there's performance / indexing reasons for making something float and integer, but then I think that should be something left up to dbType

**Comment by devinivy** [comment]
@awm33 if you have any ideas around how to deal with these issues, please weigh in!  If there's some community support behind these ideas, I think it would be worth working into 0.11.0.

**Comment by dmarcelino** [comment]
That's a reasonable suggestion, once PR #981 goes in we should start discussing those more actively. There will also be the matter of legacy... many people have been using these `types` as they are now so we'll need to balance between the ideal scenario and some sort of compromise to avoid breaking everyone's models. :)

**Comment by awm33** [comment]
@devinivy @dmarcelino  Sounds good! Me and Sam (we work together) are using Waterline in a service oriented production app. I'd love to get involved. How do you guys discuss changes / features / PRs? Just in Github issues?

**Comment by dmarcelino** [comment]
Hey @awm33, we mostly discuss these in the github issues / PRs. Feel free to submit a PR or start a discussion in an issue.

**Comment by sailsbot** [comment]
Thanks for posting, @SamProtas.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by TechupBusiness** [comment]
Hey guys, I'm a bit confused that this basic type is not implemented? EVERY application that uses any kind of money values needs a DECIMAL/DOUBLE type!? Of course I can workaround with adding database-specific type via columnType, but I hope that this is not the architecture for this "database independent"!!! ORM. :) Or is it added and just not fixed in the documentation?
